File: Base/Desktop/Allors/Client/Base/Workspace/Session.ts
namespace Allors {
    export interface ISession {
        hasChanges: boolean;

        get(id: string): any;

        create(objectTypeName: string): any;

        save(): Data.SaveRequest;
        onSaved(saveResponse: Data.SaveResponse): void;
        reset(): void;
    }

    export class Session implements ISession {
        private static idCounter = 0;

        private workspace: IWorkspace;
        private sessionObjectById: { [id: string]: ISessionObject; } = {};
        private newSessionObjectById: { [id: string]: INewSessionObject; } = {};

        constructor(workspace: IWorkspace) {
            this.workspace = workspace;
        }

        get hasChanges(): boolean {
            var hasChanges = false;
            _.forEach(this.sessionObjectById, sessionObject => {
                if (sessionObject.hasChanges) {
                    hasChanges = true;
                    return;
                }
            });

            return hasChanges;
        }

        get(id: string): any {
            if (id === undefined) {
                return undefined;
            }

            let sessionObject = this.sessionObjectById[id];
            if (sessionObject === undefined) {
                const workspaceObject = this.workspace.get(id);

                const type = Domain[workspaceObject.objectType.name];
                sessionObject = new type();
                sessionObject.session = this;
                sessionObject.workspaceObject = workspaceObject;
                sessionObject.objectType = workspaceObject.objectType;

                this.sessionObjectById[sessionObject.id] = sessionObject;
            }

            return sessionObject;
        }

        create(objectTypeName: string): any {
            const type = Domain[objectTypeName];
            const newSessionObject: INewSessionObject = new type();
            newSessionObject.session = this;
            newSessionObject.objectType = this.workspace.objectTypeByName[objectTypeName];
            newSessionObject.newId = (--Session.idCounter).toString();

            this.newSessionObjectById[newSessionObject.newId] = newSessionObject;

            return newSessionObject;
        }

        save() : Data.SaveRequest {
            var data = new Data.SaveRequest();
            data.newObjects = [];
            data.objects = [];

            if (this.newSessionObjectById) {
                _.forEach(this.newSessionObjectById, newSessionObject => {
                    var objectData = newSessionObject.saveNew();
                    if (objectData !== undefined) {
                        data.newObjects.push(objectData);
                    }
                });
            }

            _.forEach(this.sessionObjectById, sessionObject => {
                var objectData = sessionObject.save();
                if (objectData !== undefined) {
                    data.objects.push(objectData);
                }
            });

            return data;
        }
        
        onSaved(saveResponse: Data.SaveResponse): void {
            if (saveResponse.newObjects) {
                _.forEach(saveResponse.newObjects, saveResponseNewObject => {
                    var newId = saveResponseNewObject.ni;
                    var id = saveResponseNewObject.i;

                    var newSessionObject = this.newSessionObjectById[newId];

                    var loadResponse: Data.LoadResponse = {
                        userSecurityHash: "#", // This should trigger a load on next check
                        objects: [
                            {
                                i: id,
                                v: "",
                                t: newSessionObject.objectType.name,
                                roles: [],
                                methods: []
                            }
                        ]
                    }

                    delete (this.newSessionObjectById[newId]);
                    delete(newSessionObject.newId);

                    this.workspace.load(loadResponse);
                    var workspaceObject = this.workspace.get(id);
                    newSessionObject.workspaceObject = workspaceObject;

                    this.sessionObjectById[id] = newSessionObject;
                });
            }

            if (Object.getOwnPropertyNames(this.newSessionObjectById).length !== 0) {
                throw "Not all new objects received ids";
            }
        }

        reset(): void {
            _.forEach(this.newSessionObjectById, v => {
                v.reset();
            });

            _.forEach(this.sessionObjectById, v => {
                v.reset();
            });
        }
    }
}

File: Base/Workspace.Domain/Base/Workspace/Session.cs
namespace Allors.Workspace
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    using Allors.Workspace.Data;
    using Meta;

    public interface ISession
    {
        bool hasChanges { get; }

        ISessionObject get(long id);

        ISessionObject create(Class @class);

        Data.PushRequest pushRequest();

        void pushResponse(Data.PushResponse saveResponse);

        void reset();
    }

    public class Session : ISession
    {
        private static long idCounter = 0;

        private Workspace workspace;
        private Dictionary<long, SessionObject> sessionObjectById = new Dictionary<long, SessionObject>();
        private Dictionary<long, SessionObject> newSessionObjectById = new Dictionary<long, SessionObject>();

        public Session(Workspace workspace)
        {
            this.workspace = workspace;
        }

        public bool hasChanges
        {
            get
            {
                return this.newSessionObjectById.Count > 0 || this.sessionObjectById.Values.Any(v => v.hasChanges);
            }
        }

        public ISessionObject get(long id)
        {
            SessionObject sessionObject;
            if (!this.sessionObjectById.TryGetValue(id, out sessionObject))
            {
                var workspaceObject = this.workspace.get(id);

                sessionObject = this.workspace.ObjectFactory.Create(this, workspaceObject.objectType);

                sessionObject.workspaceObject = workspaceObject;
                sessionObject.objectType = workspaceObject.objectType;

                this.sessionObjectById[workspaceObject.id] = sessionObject;
            }

            return sessionObject;
        }

        public ISessionObject create(Class @class)
        {
            var newSessionObject = this.workspace.ObjectFactory.Create(this, @class);

            var newId = --Session.idCounter;
            newSessionObject.newId = newId;
            newSessionObject.objectType = @class;

            this.newSessionObjectById[newId] = newSessionObject;

            return newSessionObject;
        }

        public void reset()
        {
            foreach (var newSessionObject in this.newSessionObjectById.Values)
            {
                newSessionObject.reset();
            }

            foreach (var sessionObject in this.sessionObjectById.Values)
            {
                sessionObject.reset();
            }
        }

        public PushRequest pushRequest()
        {
            var data = new PushRequest
                           {
                               newObjects = new List<PushRequestNewObject>(),
                               objects = new List<PushRequestObject>()
                           };

            foreach (var newSessionObject in this.newSessionObjectById.Values)
            {
                var objectData = newSessionObject.saveNew();
                if (objectData != null)
                {
                    data.newObjects.Add(objectData);
                }
            }

            foreach (var sessionObject in this.sessionObjectById.Values)
            {
                var objectData = sessionObject.save();
                if (objectData != null)
                {
                    data.objects.Add(objectData);
                }
            }

            return data;
        }
        
        public void pushResponse(PushResponse pushResponse)
        {
            if (pushResponse.newObjects != null && pushResponse.newObjects.Length > 0)
            {
                foreach (var pushResponseNewObject in pushResponse.newObjects)
                {
                    var newId = long.Parse(pushResponseNewObject.ni);
                    var id = long.Parse(pushResponseNewObject.i);

                    var newSessionObject = this.newSessionObjectById[newId];

                    var loadResponse = new SyncResponse
                                           {
                                               userSecurityHash = "#",
                                               // This should trigger a load on next check
                                               objects =
                                                   new[]
                                                       {
                                                           new SyncResponseObject
                                                               {
                                                                   i = id.ToString(),
                                                                   v = "",
                                                                   t =
                                                                       newSessionObject
                                                                       .objectType.Name,
                                                                   roles = new object[0][],
                                                                   methods = new string[0][]
                                                               }
                                                       }
                                           };

                    this.newSessionObjectById.Remove(newId);
                    newSessionObject.newId = null;

                    this.workspace.sync(loadResponse);
                    var workspaceObject = this.workspace.get(id);
                    newSessionObject.workspaceObject = workspaceObject;

                    this.sessionObjectById[id] = newSessionObject;
                }
            }

            if (this.newSessionObjectById != null && this.newSessionObjectById.Count != 0)
            {
                throw new Exception("Not all new objects received ids");
            }
        }
    }
}